File: java/Weed3/src/noear/weed/cache/LocalCache.java
package noear.weed.cache;

import java.util.*;

/**
 * Created by yuety on 14-6-16.
 * 嵌入式缓存效率高；基于总量与超时双重管理(比自动超时要简单)
 *
 */
public class LocalCache implements ICacheService {
    private String _cacheKeyHead;
    private int _defaultSeconds;

    private int _max = 50000;
    private int _count = 0;

    private List<Integer> mks = new ArrayList<Integer>(); //key的顺序记录
    private HashMap<Integer, LocalCacheRecord> mcc = new HashMap<Integer, LocalCacheRecord>();   //缓存存储器

    public LocalCache(String keyHeader, int defSeconds) {
        this(keyHeader, defSeconds, 50000);
    }

    public LocalCache(String keyHeader, int defSeconds, int recordMax) {
        _cacheKeyHead = keyHeader;
        _defaultSeconds = defSeconds;
        _max = recordMax;
    }

    public void store(String key, Object obj, int seconds) {
        Integer hashKey = (_cacheKeyHead + "$" + key).hashCode();
        LocalCacheRecord val = new LocalCacheRecord(obj, seconds);

        mcc.put(hashKey, val);
        mks.add(hashKey);

        _count++;
        if (_count > _max) //总量控制
        {
            Integer k = mks.get(0);
            mcc.remove(k);
            mks.remove(0);
            _count--;
        }
    }

    public Object get(String key) {
        Integer hashKey = (_cacheKeyHead + "$" + key).hashCode();
        LocalCacheRecord val = mcc.get(hashKey);

        if (val == null)
            return null;

        if (val.time < new Date().getTime()) {
            mcc.remove(hashKey);
            mks.remove(hashKey);
            _count--;
            return null;
        } else
            return val.data;
    }

    public void remove(String key) {
        Integer hashKey = (_cacheKeyHead + "$" + key).hashCode();

        mcc.remove(hashKey);
        mks.remove(hashKey);
        _count--;
    }

    public void clear() {
        mks.clear();
        mcc.clear();
        _count = 0;
    }

    public int getDefalutSeconds() {
        return _defaultSeconds;
    }

    public String getCacheKeyHead() {
        return _cacheKeyHead;
    }
}

File: .net & mono/Weed3/Cache/LocalCache.cs
using System;
using System.Collections.Generic;

namespace Noear.Weed.Cache {
    public class LocalCache : ICacheServiceEx {
        private String _cacheKeyHead;
        private int _defaultSeconds;

        private int _max = 50000;
        private int _count = 0;

        private List<int> mks = new List<int>(); //key的顺序记录
        private Dictionary<int, LocalCacheRecord> mcc = new Dictionary<int, LocalCacheRecord>();   //缓存存储器

        public LocalCache(String keyHeader, int defSeconds) : this(keyHeader, defSeconds, 50000) {

        }

        public LocalCache(String keyHeader, int defSeconds, int recordMax) {
            _cacheKeyHead = keyHeader;
            _defaultSeconds = defSeconds;
            _max = recordMax;
        }

        public void store(String key, Object obj, int seconds) {
            int hashKey = (_cacheKeyHead + "$" + key).GetHashCode();
            LocalCacheRecord val = new LocalCacheRecord(hashKey, obj, seconds);
            mcc.Add(hashKey, val);
            mks.Add(hashKey);

            _count++;
            if (_count > _max) //总量控制
            {
                int k = mks[0];
                mcc.Remove(k);
                mks.Remove(0);
                _count--;
            }
        }

        public Object get(String key) {
            int hashKey = (_cacheKeyHead + "$" + key).GetHashCode();

            LocalCacheRecord val = null;
            if (mcc.ContainsKey(hashKey))
                val = mcc[hashKey];

            if (val == null)
                return null;

            if (val.time < new DateTime().Ticks) {
                mcc.Remove(hashKey);
                mks.Remove(hashKey);
                _count--;
                return null;
            }
            else
                return val.data;
        }

        public void remove(String key) {
            int hashKey = (_cacheKeyHead + "$" + key).GetHashCode();

            mcc.Remove(hashKey);
            mks.Remove(hashKey);
            _count--;
        }

        public int getDefalutSeconds() {
            return _defaultSeconds;
        }

        public String getCacheKeyHead() {
            return _cacheKeyHead;
        }


        //===================
        //
        public CacheTags tags() {
            return new CacheTags(this);
        }

        public void clear(string tag) {
            tags().clear(tag);
        }

        public void update<T>(string tag, Func<T, T> setter) where T : class {
            tags().update<T>(tag, setter);
        }
    }
}
